content-security-policy/script-src/script-src-strict_dynamic_parser_inserted.html

```diff
@@ -0,0 +1,193 @@
+            window.addEventListener('message', t.step_func(function(e) {
+                if (e.data === 'documentWrite-deferred-async') {
+                    assert_unreached('Parser-inserted deferred async script via `document.write` without a correct nonce is not allowed with `strict-dynamic`.');
+                }
+            }));
+            window.addEventListener('securitypolicyviolation', t.step_func_done(function(violation) {
+                if (violation.target.id !== 'documentWrite-deferred-async') {
+                    return;
+                }
+                assert_equals(violation.effectiveDirective, 'script-src');
+            }));
+
+            document.write('<scr' + 'ipt defer async id="documentWrite-deferred-async" src="simpleSourcedScript.js"></scr' + 'ipt>');
+        }, 'Parser-inserted deferred async script via `document.write` without a correct nonce is not allowed with `strict-dynamic`.');
+    </script>
+
+    <script nonce='dummy'>
+        async_test(function(t) {
+            window.addEventListener('message', t.step_func(function(e) {
+                if (e.data === 'documentWriteln-deferred-async') {
+                    assert_unreached('Parser-inserted deferred async script via `document.writeln` without a correct nonce is not allowed with `strict-dynamic`.');
+                }
+            }));
+            window.addEventListener('securitypolicyviolation', t.step_func_done(function(violation) {
+                if (violation.target.id !== 'documentWriteln-deferred-async') {
+                    return;
+                }
+                assert_equals(violation.effectiveDirective, 'script-src');
+            }));
+
+            document.writeln('<scr' + 'ipt defer async id="documentWriteln-deferred-async " src="simpleSourcedScript.js"></scr' + 'ipt>');
+        }, 'Parser-inserted deferred async script via `document.writeln` without a correct nonce is not allowed with `strict-dynamic`.');
+    </script>
+
+    <script nonce='dummy'>
+        var innerHTMLScriptRan = false;
+        async_test(function(t) {
+            window.addEventListener('securitypolicyviolation', t.step_func_done(function(violation) {
+                if (violation.target.id !== 'innerHTML') {
+                    return;
+                }
+                assert_false(innerHTMLScriptRan);
+                assert_equals(violation.effectiveDirective, 'script-src');
+            }));
+            var e = document.createElement('div');
+            e.innerHTML = "<img id='innerHTML' src='/nonexisting.jpg' onerror='innerHTMLScriptRan = true;' style='display:none'>";
+            document.body.appendChild(e);
+        }, 'Script injected via `innerHTML` is not allowed with `strict-dynamic`.');
+    </script>
+
+    <script nonce='dummy'>
+        var insertAdjacentHTMLScriptRan = false;
+        async_test(function(t) {
+            window.addEventListener('securitypolicyviolation', t.step_func_done(function(violation) {
```

Comment:
I think these should all be `step_func()` rather than `step_func_done()`. The latter will complete the test when _any_ `securitypolicyviolation` event is fired (even if you return early without doing any assertions). Switching to `step_func()` and calling `t.done()` manually after your assertions should mean that you're actually testing each block individually.